Все исключения в Kotlin являются наследниками класса **`Throwable`**. У каждого исключения есть сообщение, трассировка стека и (опционально) причина, по которой это исключение вероятно было вызвано.

Для того чтобы возбудить исключение явным образом, используйте оператор **`throw`**.

```kotlin
throw Exception("Hi There!")
```

Чтобы перехватить исключение, используйте выражение **`try`**...**`catch`**.

```kotlin
try { 
	// some code 
} catch (e: SomeException) { 
	// handler
} finally { 
	// optional finally block 
}
```

В коде может быть любое количество блоков `catch` (такие блоки могут и вовсе отсутствовать). Блоки `finally` могут быть опущены. Однако, должен быть использован как минимум один блок `catch` или `finally`.

### Try - это выражение

`try` является выражением, что означает, что оно может иметь возвращаемое значение.

```kotlin
val a: Int? = try { input.toInt() } catch (e: NumberFormatException) { null }
```

Возвращаемым значением будет либо последнее выражение в блоке `try`, либо последнее выражение в блоке `catch` (или блоках). Содержимое `finally` блока никак не повлияет на результат `try`-выражения.

`throw` в Kotlin является выражением, поэтому есть возможность использовать его, например, в качестве части Elvis-выражения:

```kotlin
val s = person.name ?: throw IllegalArgumentException("Name required")
```

Типом выражения `throw` является специальный тип под названием `Nothing`. У этого типа нет никаких значений, он используется для того, чтобы обозначить те участки кода, которые могут быть не достигнуты никогда. В своём коде вы можете использовать `Nothing` для того, чтобы отметить функцию, чей результат никогда не будет возвращён.

```kotlin
fun fail(message: String): Nothing {
    throw IllegalArgumentException(message)
}
```

При вызове такой функции компилятор будет в курсе, что исполнения кода далее не последует:

```kotlin
val s = person.name ?: fail("Name required")
println(s) // известно, что переменная 's' проинициализирована к этому моменту
```

Вы также можете столкнуться с этим типом при работе с выводом типов. Nullable-вариант этого типа `Nothing?` имеет ровно одно возможное значение - `null`. Если вы используете `null` для инициализации значения предполагаемого типа, и нет никакой другой информации, которую можно использовать для определения более конкретного типа, компилятор определит тип `Nothing?`.

```kotlin
val x = null           // у 'x' тип `Nothing?`
val l = listOf(null)   // у 'l' тип `List<Nothing?>
```

[[Проверяемые исключения]]